### DISCUSSION TITLE
Windows fatal exception: access violation

### DISCUSSION THREAD

**Discussion by Thoufak** [pyinstaller/pyinstaller#7287] [Help]
Sometimes a compiled version of my program will crash with this exception. It would look something like this:

<img width="400" alt="Screenshot 2022-11-19 at 00 42 43" src="https://user-images.githubusercontent.com/40132582/204361911-b6d7eb12-0796-4468-b8c9-490c23a3e890.png">

I’ve observed some notable features about when and how this happens, and I will list them below. Please note that I’m not sure if this is a PyInstaller bug, but I hope you can help me find out. Also note that I’m using pyarmor. I’ve been using pyarmor + PyInstaller for over a year for this project, but I started seeing this problem only recently.

## Rebuilding helps
If I see that the resulting exe has this issue, I can build it again using exactly the same source code and command. Nothing seems to be changing from my side. And this will produce a build which doesn’t have the crash, or it will now appear in a different place (see the next point).

## It crashes on different lines.
If a specific build has this crash, and I don’t change any input parameters to my program, it will crash on the same line every time I start it (I mainly looked at the main thread stacktrace). In other words, this crash is very well reproducible within the same build. However, as I said, if I make a new build, this error will either be gone (most likely) or appear in a completely different place. I couldn’t find any meaningful similarities between the different lines of code I’ve seen my program crashing at.

Notice how I said "*and I don’t change any input parameters to my program*". I’ve once observed a crash inside a ‘for’ loop which is used to load lines from a text file. I managed to find the exact number of lines needed to crash my program. If the number of lines was below, it wouldn’t crash on that loop (but it probably will crash a bit later. I don’t remember this detail but I can check). It looks as though there's a defined number of "operations" that it does before crashing. 

Also, it seems that this crash can happen in any place in the code. Sometimes it’s right at the start of the program, but sometimes you have to use it for a while to see the crash.

## Visual Studio debugger
I’ve tried attaching a debugger from Visual Studio. I don’t think it produced any useful information, but I will attach the log anyway.
[VSCode debugger.txt](https://github.com/pyinstaller/pyinstaller/files/10106772/VSCode.debugger.txt)

## More tracebacks examples
These are from different builds.

<img width="400" alt="Screenshot 2022-11-10 at 02 57 35" src="https://user-images.githubusercontent.com/40132582/204363021-5404f096-7403-41c4-a709-65a9c14016d3.png">
<img width="400" alt="Screenshot 2022-09-30 at 13 10 12" src="https://user-images.githubusercontent.com/40132582/204363032-9fc2a65b-c862-441b-aa14-b13847c6a3a2.png">

## Versions
**PyInstaller**: 4.9
**Pyarmor**: 7.4.3
**Python**: 3.9.5
**EXE built on**: Windows 11 Home ARM64 (this is a machine from Parallels Desktop running on Mac M1 Pro)

**Comment by bwoodsend** [comment]
_access violation_ on Windows translates to _some broken code written in C tried to read a bit of memory it doesn't own_. These issues can pass under the radar if said piece of C code reads memory that it does own but isn't the piece of memory it intended to read and because of the fairly deterministic nature of the OS's memory allocation system, whether the broken code reads out of process memory or just the wrong memory will go through phases of consistently doing one or the other if you just keep restarting your program without doing something else to in between to disrupt the memory cycle. This, I guess, is the only reason running a rebuild changes anything.

PyInstaller does contain C code but it's all in launching/cleanup - once you're application is going PyInstaller's C code is out of the way. pyarmour on the other hand is doing some rather scary bytecode decryption in C every time you enter a new function. Your vslog shows other non pure Python libraries in the mix too. So long story short, I'd put PyInstaller pretty low down on my list of suspects here.

**Reply by Thoufak** [reply]
Thank you for your answer.

**Reply by Thoufak** [reply]
Can you suggest any specific ways to further investigate this?

**Reply by bwoodsend** [reply]
Nothing novel. Just keep removing libraries and unknowns like pyarmour until the crashes cease.

**Reply by rokm** [reply]
> Can you suggest any specific ways to further investigate this?

The first thing I'd try is disabling PyArmor and see if the crashes still occur or not.
